deepchecks/tabular/utils/task_inference.py

```diff
@@ -39,65 +40,57 @@ def get_possible_classes(model: Optional[BasicModel], train_dataset: 'tabular.Da
-            get_logger().warning('Model classes attribute does not contain all observed labels in train and test data.')
-        observed_labels += list(model.classes_)
+    if train_dataset and train_dataset.label_type is not None:
+        task_type = train_dataset.label_type
+    elif model_classes:
+        task_type = infer_by_class_number(len(model_classes))
+    elif is_categorical(observed_labels, max_categorical_ratio=0.05):
+        num_classes = len(observed_labels.dropna().unique())
+        task_type = infer_by_class_number(num_classes)
+        if infer_dtype(observed_labels) == 'integer' and train_dataset and train_dataset.label_type is None:
+            get_logger().warning(
+                'Due to the small number of unique labels task type was inferred as classification in spite of '
+                'the label column is of type integer. '
+                'Initialize your Dataset with either label_type=\"multiclass\" or '
+                'label_type=\"regression\" to resolve this warning.')
+    else:
+        task_type = TaskType.REGRESSION
 
-    if model is not None:  # classification model without classes_ attribute
-        observed_labels += convert_into_flat_list(model.predict(train_dataset.features_columns))
-        if test_dataset is not None:
-            observed_labels += convert_into_flat_list(model.predict(test_dataset.features_columns))
-        if hasattr(model, 'predict_proba'):  # This means it's a classification model
-            return sorted(pd.Series(observed_labels).dropna().unique())
-    label_series = pd.Series(observed_labels)
-    if is_categorical(label_series, max_categorical_ratio=0.05) or force_classification:
-        return sorted(label_series.dropna().unique())
-    else:  # no predict_proba method + not categorical column (regression)
-        return None
+    if task_type in [TaskType.BINARY, TaskType.MULTICLASS]:
+        return task_type, sorted(observed_labels.dropna().unique()), model_classes
+    else:
+        return task_type, None, None
 
 
-def infer_task_type(model: Optional[BasicModel], train_dataset: 'tabular.Dataset',
-                    test_dataset: Optional['tabular.Dataset'] = None) -> TaskType:
-    """Infer the task type based on get_possible_classes.
-
-    Parameters
-    ----------
-    model : BasicModel
-        Model object used in task
-    train_dataset : 'tabular.Dataset'
-        Train Dataset of task
-    test_dataset : Optional['tabular.Dataset'], default = None
-        Test Dataset of task
-    Returns
-    -------
-    TaskType
-        The type of the Task
-    """
-    classes = get_possible_classes(model, train_dataset, test_dataset)
-    if classes is None:
-        return TaskType.REGRESSION
-    elif len(classes) == 2:
-        return TaskType.BINARY
-    else:
-        return TaskType.MULTICLASS
+def infer_by_class_number(num_classes):
+    if num_classes == 0:
+        raise ValidationError('Found zero number of classes')
```

Comment:
```suggestion
        raise ValidationError('Found no classes')
```
How can that happen? Only nulls? Maybe in that case also the user should pass model_classes

Reply:
I don't see a reason this will happen, but again I prefer to have it fail loudly in some weird edge use case. we also need to add validation that we don't get empty model_classes list